opensearchapi/api.bulk.go

```diff
@@ -116,8 +116,16 @@ func (r BulkRequest) Do(ctx context.Context, transport Transport) (*Response, er
 		params["routing"] = r.Routing
 	}
 
-	if len(r.Source) > 0 {
-		params["_source"] = strings.Join(r.Source, ",")
+	if source, ok := r.Source.(bool); ok {
+		params["_source"] = strconv.FormatBool(source)
+	}
+
+	if source, ok := r.Source.(string); ok && source != "" {
+		params["_source"] = source
+	}
+
+	if sources, ok := r.Source.([]string); ok && len(sources) > 0 {
+		params["_source"] = strings.Join(sources, ",")
 	}
```

Comment:
Yes, in the draft, I used the `switch` constructor, but it did not fit with the entire structure that was in the file, that is, from above it is also possible to replace all `if` constructors, but since it was originally laid down, I did not change the writing style and left if the constructor.  In practice, this does not change anything, but aesthetically it will look beautiful, do you agree?

Reply:
Using switch makes a difference compared to if **without** else. If you change it to if + else the performance would not improve.

Personally I prefer using switch.

Reply:
Notice I didn't say a word about performance, I was talking about aesthetics and practicality. When using the `if / else` constructor, practically nothing will change, but it will look much better in resonance with the rest of the code. if I'm wrong, please give an example where exactly this constructor is worse than `switch`, except for a subjective opinion

Reply:
I am not sure I can follow you with "aesthetics and practicality". Currently the if without the else is bad code as it won't skip checking once it found a match. Therefore we have two option. Either using if/else or switch.
The current if check for parameters are imo ugly and produce a lot of duplicate code, so I am currently thinking about a better way to handle those parameters. Therefore I would not focus on how the parameters are currently handled with if's in reference to  "aesthetics".

Reply:
Ok, let's assume that user pass `int` to `_source` field what would be the scenario of checking:

`If / else` condition:
1. Check if type `bool` condition, not matched
2. Check if type `string` condition, not matched
3. Check if type `[]string` condition, not matched
4. Do nothing with this field

`switch` condition:
1. Check if type `bool` condition, not matched
2. Check if type `string` condition, not matched
3. Check if type `[]string` condition, not matched
4. Do nothing with this field
 
What is practically difference of this? 

`Currently the if without the else is bad code as it won't skip checking once it found a match`
Do you agree that in switch scenario we do same? What can you pass to default case in switch?

_Aesthetics._ I mean that rest of code also written with `if` constructor, so why we should use `switch` constructor in the middle of the code, if there is no difference in logic? I agree if we consider the block `switch` as a separate piece, it fits to this conditions pretty well, but if you look at the whole code, is this at least, inappropriate?

`if` constructor much clear to understand, even beginners can identify what is going on. 
 
`The current if check for parameters are imo ugly and produce a lot of duplicate code`
Ok, let's refactor less of the code to `switch` operator, since 70% of whole project with `if` constructors

I think this just subjective opinion and it's time to connect others for the review

@dblock @VijayanB @VachaShah

Reply:
As already sad, the if without else is bad code because it continues checking once a match is found. Assuming the client passes a `bool`.
Current Code:
1. Check the `bool` condition, match
2. Check the `string` condition, no match
3. Check the `[]string` condition, no match
4. Continue with other Code

If/else or switch:
1. Check the `bool` condition, match
2. Continue with other Code

> What is practically difference of this?

We save runtime as it don´t need to check the other cases.

> Do you agree that in switch scenario we do same? What can you pass to default case in switch?

No and we don't need a default for switch.

> I mean that rest of code also written with if constructor, so why we should use switch constructor in the middle of the code, if there is no difference in logic?

Just to clarify. The whole discussion about  _aesthetics_ is to decide between if/else and switch. The current code is no option. Depending on how strict you see it, using if/else is also a "stile" brake compared to the rest of the code.
However it would be okay to use if/else instead of switch.
The other code is written with if constructors as this is the simplest way of doing it. Using switch would require looping over a struct which makes thinks more complicated but reduces duplicate code. 

To complete, yes using switch is my opinion as it is untypical to use else clauses in golang.

Reply:
Ok, this argument now seems valid to me, but I still think that the switch is inappropriate here, so I propose this constructor

```
if source, ok := r.Source.(bool); ok {
  params["_source"] = strconv.FormatBool(source)
} else if source, ok := r.Source.(string); ok && source != "" {
  params["_source"] = source
} else if sources, ok := r.Source.([]string); ok && len(sources) > 0 {
  params["_source"] = strings.Join(sources, ",")
}
```

Reply:
Your proposed if/else is technically okay and is valid.

Personally, I find it easier to read and understand the suggested switch compared to the if/else even though the switch uses more lines of code.

As this is my personal opinion, I would like to get @dblock and/or @VachaShah opinion on this.